File: server/models/goal.js
/**
 * Created by juwoong on 2015. 8. 19..
 */
var mongoose = require('mongoose');
var Schema = mongoose.Schema;

var Result = new Schema({
    date : Date,
    percent : Number,
    result : Number
});

var Diary = new Schema({
    userId : {type: Schema.ObjectId, ref:'user'},
    goalId : {type: Schema.ObjectId, res:'goal'},
    date : Date,
    content : String
});


var Farewell = new Schema({
    isMonthly       : Boolean,
    name            : String,
    generateAt      : Date,
    result          : [Number],
    total           : Number,
    totalComment    : {
        main        : String,
        sub         : String
    },
    average         : Number,
    count           : Number,
    fail            : [Number],
    endurance       : {
        main        : String,
        sub         : String,
        percent     : Number
    }
});

var UserInGoal = new Schema({
    Id : Schema.Types.ObjectId,
    name : { type : String, max : 10 },
    profileImage : { type : String, length : 64 },
    target : {
        score : Number,
        type : {type : String}
    },
    startAt : Date,
    result : [Result],
    diaries : [{type: Schema.ObjectId, res:'diary'}],
    farewell : Farewell,
});

var Goal = new Schema({
    name : { type : String, max : 100 },
    tags : [String],
    startDate : Date,
    endDate : Date,
    category : String,
    threeLegged : Boolean,
    users : [UserInGoal],
});

module.exports.Diary = mongoose.model('diary', Diary);
module.exports.Goal = mongoose.model('goal', Goal);
module.exports.UserInGoal = UserInGoal;
module.exports.Result = Result;

File: worker/structs/goal.go
package structs

import (
	"gopkg.in/mgo.v2/bson"
	"time"
)

type Target struct {
	Score int    `bson:"score"`
	Type  string `bson:"type"`
}

type Result struct {
	Date    time.Time `bson:"date"`
	Reason  int       `bson:"result"`
	Percent int       `bson:"percent"`
}

type UserInGoal struct {
	Id           bson.ObjectId `bson:"Id"`
	Name         string        `bson:"name"`
	ProfileImage string        `bson:"profileImage"`
	Target       Target        `bson:"target"`
	StartAt      time.Time     `bson:"startAt"`
	Result       []Result      `bson:"result"`
}

type Goal struct {
	Id          bson.ObjectId `bson:"_id,omitempty" json:"id"`
	Name        string        `bson:"name"`
	StartDate   time.Time     `bson:"startDate"`
	EndDate     time.Time     `bson:"endDate"`
	Category    string        `bson:"category"`
	ThreeLegged bool          `bson:"threeLegged"`
	Users       []UserInGoal  `bson:"users"`
}
